students/ThanhNhan/lesson09/mail5.py

```diff
@@ -0,0 +1,197 @@
+
+    def exist_list(self):
+        return [person_name for person_name in self.data_collection]
+
+    
+    def find_donor(self):
+        
+        for x in self.data_collection.keys():
+            print (x, "test")
+            if (x == self.name):
+                return True
+        return False
+
+    def create_report(self):
+        report_donor = []
+        print(self.data_collection, " report")
+
+        for key, values in self.data_collection.items():    
+            name = "{}".format(key)
+            total_cost = sum(values)
+            n_time = len(values)
+            average = total_cost/n_time
+            report_donor.append([name, round(total_cost, 2), n_time, round(average, 2)])
+
+        report_donor = sorted(report_donor, key=self._sort, reverse=False)
+
+        print("{:20}|  {:<10s}|  {:<10s}|  {: <10s} ".format(*self.title))
+        print("".join(["-"] * 62))
+        for x in range (len(report_donor)):
+            print("{:20} $ {:10.2f} {:>12d} $ {:>12.2f}".format(*report_donor[x]))
+
+        return report_donor
+
+    @staticmethod
+    def _sort(row_report):
+        #each row record look like
+        dl_name = row_report[0].split(' ')[1]
+        if (len(dl_name) > 1):
+            return dl_name
+        else:
+            return row_report[0].split(' ')[2]
+
+    def ck_dir(self):
+        if not os.path.isdir(self.DIR_NAME):
+            os.mkdir(self.DIR_NAME)
+            return False
+        return True
+
+    def ss_letterall(self):
+    #     """
+    #     This method write letter to disk as txt file
+
+    #     params:  none
+    #     will check and generate directory
+    #     will write all txt file in the directory
+    #     """
+
+        #writes each letter to disk as txt file
+        #open ("./thankyou.txt", "w")
+        # tempfile.gettempdir() 
+
+        for key, values in self.data_collection.items():  
+            d = Donor(key)
+            letter = d.generate_letter(values)
+            filename = key + ".txt"
+            self.ck_dir()    
+            try:
+                print(filename) 
+                filename = os.path.join(self.DIR_NAME, filename)
+                open(filename, 'w').write(letter)
+            except FileNotFoundError:
+                print("couldn't open missing.txt")
+            finally:
+                pass
+    
+
+    def add_donor(self, Donor):
+        if self.find_donor() is True:
+            self.data_collection[Donor.name].append(Donor.donations[0])    
+        else:
+            self.data_collection[Donor.name] = Donor.donations
+
+
+
+class Donor(DonorCollection):
+
+    def __init__(self, name, donations=None):  
+        """
+        Create Donor object
+        param name: the full name of the donor
+        param donations: the donations 
+        """
+
+        self.name = name
+        if donations is None:
+            self.donations = []
+        else:
+            self.donations = [donations]
```

Comment:
this will crate a list with donations in it -- which is what you want if there is a single donation passed in. Casreful with naming -- calling it "donations" makes me expect it to be more than one!